ucb-rit/coldfront
# Support renewal requests and new project requests across allocation periods

Refs #295

**Changes**
- Summary
	- Added support for requesting to renew allowances (currently FCAs only) or create new projects for the next allowance year, available only in the month preceding the start of the year, using the existing forms.
	- Updated data modeling and application logic so that requests can be in an approved, but not yet processed, state.
		- For example, a request for the next allowance year should not be processed until the year begins, but may be approved in advance.
		- Conversely, a request for the current allowance year can be processed immediately after approval.
	- Added a scheme for scheduling processing of approved requests at the correct times without human intervention.
	- Added and updated tests.
- Models
	- Added a new `ProjectAllocationRequestStatusChoice`: "Approved - Scheduled".
	- Added `allocation_period`, `request_time`, `approval_time`, and `completion_time` fields to `SavioProjectAllocationRequest`.
- Signals
	- When an `AllocationPeriod` is created, a task for running the management command `start_allocation_period` is scheduled for the beginning of that period.
    - When a period is updated/deleted, already-scheduled tasks are updated/deleted accordingly.
- Forms
	- Included `AllocationPeriod` as an option in the forms for creating a new project, renewing a PI's allocation, and renewing a PI's allocation under a particular project.
		- The selectable periods are based on the current time and feature flags.
	- Updated selectable PIs such that eligibility only considers requests under the requested `AllocationPeriod`.
- Templates
	- Added context processors to include two context variables across all templates: `CURRENT_ALLOWANCE_YEAR_ALLOCATION_PERIOD` and `DISPLAY_TIME_ZONE`.
	- Added an alert (conditionally displayed based on feature flags) to the landing views for new project requests and renewal requests to notify users that they may make requests under the next allowance year.
	- Updated text across the site and in emails to account for requests that were approved, but not yet processed.
	- Stopped displaying pending new project requests on the landing view because the Request Hub makes them accessible.
- Views
	- Updated request list views:
		- Color-coded requests to indicate whether their respective `AllocationPeriods` have started, allowing administrators to prioritize them.
		- Updated "Pending"/"Completed" designations so that "Approved" and "Approved - Scheduled" requests are considered "Completed", since processing will be done automatically.
	- Removed hard-coded `AllocationPeriods` and filtered requests by `AllocationPeriod`.
- Management Commands
    - `create_allocation_period`: Create hard-coded `AllocationPeriods`. This will need to be updated periodically to include new periods as they approach.
    - `set_new_project_request_allocation_periods`: Fill in missing `AllocationPeriods` for `SavioProjectAllocationRequests`.
    - `set_new_project_request_times`: Fill in missing times for `SavioProjectAllocationRequests` based on historical objects.
    - `start_allocation_period` (Incomplete): Begin a particular `AllocationPeriod` (designed to run on its `start_date`).
	    - FCAs/PCAs: Deactivate all existing FCAs/PCAs, and then process all "Approved" renewal requests (FCAs only for now) and all "Approved - Scheduled" new project requests under the period.
		- ICAs: Deactivate all ICAs that have ended, and process all "Approved - Scheduled" new project requests under the period.
- Configuration
	- Added a setting `DISPLAY_TIME_ZONE` to avoid hard-coded time zones.
    - Added a feature flag `ALLOCATION_RENEWAL_FOR_NEXT_PERIOD_REQUESTABLE` that determines whether new project requests and renewal requests can be made for the next allowance year.
    - Added a custom `django-flags` condition `during month` that checks whether the current month is a particular month, for use in determining whether renewals for the next period can be requested.
    - Added hard-coded conditions for enabling flags.
        - Note: The values in the conditions should be set by Ansible.
- Utilities
	- Filled in logic for previously-empty approval runners, so that requests can be approved and processed separately.
	- Updated the calculation of service units to allocate to always be based on the request time, rather than the approval time, relative to the `AllocationPeriod` the request is under.
- Code Quality
    - Broke up large forms, views, utilities, etc. files into separate modules.

**How to Test**
- Review the code changes and add comments as needed.
- Ensure that the test suite passes.
- Deployment Steps
    - Re-run the Ansible playbook.
    - Rename the `AllocationPeriod` "AY21-22" to "Allowance Year 2021 - 2022" in the shell:
    ```
    from coldfront.core.allocation.models import AllocationPeriod
    allocation_period = AllocationPeriod.objects.get(name="AY21-22")
    allocation_period.name = "Allowance Year 2021 - 2022"
    allocation_period.save()
    ```
    - Run management commands.
        - `python manage.py create_allocation_periods`
        - `python manage.py set_new_project_request_allocation_periods auto`
        - `python manage.py set_new_project_request_times`
    - In the Django admin interface (`/admin/flags/flagstate/`), delete manually-created flag states.

**Notes**
- This pull request largely supports the request-side of the workflow. Support for the approval/processing-side will be required before the allowance year begins, including:
    - A command for mass-approving, but not processing, requests,
    - The completion of `start_allocation_period`, and
    - Updates to `AllocationRenewalProcessingRunner` to handle edge cases introduced by support for multiple `AllocationPeriods`.
- This pull request does not support renewals of other allocation types (ICAs, PCAs), although the logic is generalized to support them.
- #388 must be merged before the start of the next allowance year so that jobs from a previous `AllocationPeriod` do not affect current SU balances.
